packages/remix-dev/compiler/plugins/cssModulesPlugin.ts

```diff
@@ -0,0 +1,362 @@
+
+--------------------------------------------------------------------------------
+`)
+            );
+          }
+        } catch (err) {
+          console.error(`
+--------------------------------------------------------------------------------
+
+A CSS Modules file was imported, but the required \`@remix-run/css-modules\`
+dependency was not found.
+
+Install the dependency by running the following command, then restart your app.
+
+    npm install @remix-run/css-modules
+
+--------------------------------------------------------------------------------
+`);
+          throw err;
+        }
+
+        let path = getResolvedFilePath(args);
+        return {
+          pluginName: browserPluginName,
+          namespace: browserPluginNamespace,
+          path,
+          sideEffects: false,
+        };
+      });
+
+      build.onLoad(
+        { filter: suffixMatcher, namespace: browserPluginNamespace },
+        async (args) => {
+          try {
+            let processed = await processCss({
+              config,
+              filePath: args.path,
+            });
+            let json = processed?.json || {};
+            return {
+              contents: JSON.stringify(json),
+              loader: "json",
+            };
+          } catch (err: any) {
+            return {
+              errors: [{ text: err.message }],
+            };
+          }
+        }
+      );
+    },
+  };
+}
+
+/**
+ * This plugin is for the server build. It doesn't actually process the CSS, but
+ * it  and waits for the assets manifest to resolve module imports.
+ */
+export function cssModulesServerPlugin(
+  config: RemixConfig,
+  assetsManifestPromiseRef: AssetsManifestPromiseRef
+): esbuild.Plugin {
+  return {
+    name: serverPluginName,
+    async setup(build) {
+      build.onResolve({ filter: suffixMatcher }, (args) => {
+        return {
+          pluginName: serverPluginName,
+          namespace: serverPluginNamespace,
+          path: getResolvedFilePath(args),
+          sideEffects: false,
+        };
+      });
+
+      build.onLoad(
+        { filter: suffixMatcher, namespace: serverPluginNamespace },
+        async (args) => {
+          let res = await assetsManifestPromiseRef.current;
+          let json = res?.cssModules?.moduleMap[args.path]?.json || {};
+          return {
+            contents: JSON.stringify(json),
+            loader: "json",
+          };
+        }
+      );
+    },
+  };
+}
+
+export async function processCss({
+  config,
+  filePath,
+}: {
+  config: RemixConfig;
+  filePath: string;
+}): Promise<CssModuleFileContents> {
+  let classPrefix =
+    path.basename(filePath, path.extname(filePath)).replace(/\./g, "-") + "__";
+  let source = await fse.readFile(filePath);
+
+  let res = parcelTransform({
+    filename: path.relative(config.appDirectory, filePath),
+    code: source,
+    cssModules: true,
+    minify: process.env.NODE_ENV === "production",
+    analyzeDependencies: true,
```

Comment:
@chaance @devongovett when I use the `url()` function, the value in the `url()` gets hashed but never replaced again with the final result.

**input**
```
.container {
  background-image: url(https://images.unsplash.com/photo-1593288942460-e321b92a6cde)
}
```

**output**
```
.layout-module___9D6bia_container {
    background-image: url(mHxWpW);
}
```

This is because `analyzeDependencies` is set to `true` here but there is no process to replace those hashes with the actual results. [more info](https://github.com/parcel-bundler/parcel-css/blob/master/node/index.d.ts#L20-L26)

Does it need to be enabled? if so, we need to add a process to replace the hashes with the final urls again.